contracts/VestingManager.sol

```diff
@@ -0,0 +1,212 @@
+
+        (bool success, ) = owner().call{value: amount}('');
+        require(success, 'Failed to send Ether');
+    }
+
+    /**
+     * @dev Withdraws remaining unlocked tokens across offerings.
+     * @param _lockup period in quarters that tokens remain blocked
+     * @param _vesting percentage of total tokens released each quarter
+     * @return created schema ID
+     */
+    function createVestingSchema(uint8 _lockup, uint16 _vesting) external onlyOwner returns (uint8) {
+        // Validate that we are not overflowing uint8
+        uint8 newSchemaId = schemaId;
+        schemaId++;
+        require(_vesting > 0 && _vesting <= 10000, 'Vesting % should be withing 0 and 10000 (2 decimal floating point)');
+        // Store new schema and emit event
+        VestingSchemas[newSchemaId] = VestingSchema(_lockup, _vesting);
+        emit SchemaCreated(newSchemaId, _lockup, _vesting);
+
+        return newSchemaId;
+    }
+
+    /* Public functions */
+
+    /**
+     * @dev Withdraws remaining unlocked tokens across offerings.
+     * @param _to address of the recipient of the tokens
+     * @param _amount of tokens to be deposited into the contract
+     * @param _schemaId vesting schema used in the deposit
+     */
+    function deposit(
+        address _to,
+        uint256 _amount,
+        uint8 _schemaId
+    ) external {
+        // Validate input parameters
+        require(_to != address(0), 'Cannot deposit to address 0');
+        require(_amount > 0, 'Value must be positive');
+
+        // Validate that vesting schema exists by checking against unaccepted value
+        VestingSchema memory schema = VestingSchemas[_schemaId];
+        require(schema.vesting != 0, 'Vesting schema does not exist');
+
+        // Validate that spender is allowed to operate with OID token
+        require(oidToken.allowance(msg.sender, address(this)) >= _amount, 'Not enough allowance');
+        // Transfer OID token to locker contract
+        oidToken.safeTransferFrom(msg.sender, address(this), _amount);
+        // Add new vesting storage to user holdings
+        Holdings[_to].push(Holding(_amount, 0, _schemaId));
+
+        emit Deposit(_to, _amount, _schemaId);
+    }
+
+    /**
+     * @dev Withdraws remaining unlocked tokens from address.
+     * @param _from the users address
+     */
+    function claim(address _from) external referenceInitiated {
+        Holding[] storage userHoldings = Holdings[_from];
+        require(userHoldings.length > 0, 'User does not have any holdings');
+
+        uint256 availableBalance = 0;
+        for (uint256 i = 0; i < userHoldings.length; i++) {
+            uint256 unlockedBalance = _getAvailableBalance(userHoldings[i]);
+            if (unlockedBalance > 0) {
+                userHoldings[i].releasedAmount += unlockedBalance;
+                availableBalance += unlockedBalance;
+            }
+        }
+        require(availableBalance > 0, 'There are no tokens available to claim');
```

Comment:
this require will create an error if you call the function when there is nothing to claim, wouldn't it be better to simply do nothing and emit a withdraw event with 0 as `availableBalance` ?

Reply:
imo it would be cleaner in terms of user experience

Reply:
If we revert the transaction we might reduce the amount of gas used, if we continue with execution the whole amount is used

Reply:
not if you put a if(availableBalance > 0) before calling the transferFrom, then the 0 transaction would just bypass the transferFrom part and emit an event with value 0

Reply:
both approaches are ok, just that people don't like to see transaction errors, they prefer a useless transaction than an error i think